### ISSUE TITLE
Allow to use Node.js crypto library in the sandbox

### ISSUE THREAD

**Issue by JoseExposito** [a85/POSTMan-Chrome-Extension#3646]
## App Details:

```
Postman for Mac
Version 5.3.0
OS X 10.11.6 / x64
```

## Feature Request:

This is not a bug report, just a feature request.

It'd be nice if [Node's crypto module](https://nodejs.org/api/crypto.html) could be accessible from the sandbox.

I'm trying to generate a ES256 JWT token before each request in my collection but CryptoJS does not implement the ECDSA algorithm.
Also, crypto implements much more functionality that CryptoJS and I'm sure there are many use cases where this feature will be useful.

Thanks a lot for your excellent work on Postman 😄

**Comment by harryi3t** [comment]
@JoseExposito I am glad that you like our product so much :) 
Adding support for crypto is already on our roadmap. Will keep you posted.

Thank you for your patience.

**Comment by frankiedrake** [comment]
Also waiting for crypto :D Thanks!

**Comment by EskoCruz** [comment]
Same 😃

**Comment by pburgr** [comment]
mee too

**Comment by davidschwab** [comment]
@harryi3t @kamalaknn Postman/Newman is an awesome platform, but this too is a blocker for RSA support. Exposing the native node.js crypto module would solve, any ETA?

FYI - [this issue looks similar](https://github.com/postmanlabs/postman-app-support/issues/2678).

Update: In case anyone else can do the same, I worked around this by creating an external microservice to create the needed signed JWT using the RS256 algorithm.

**Comment by kingwrcy** [comment]
+1,please support RSA  encrypt/decrypt/sign/verify。

**Comment by brugr9** [comment]
+1: `crypto.createHmac('sha1', pm.apiToken)`

**Comment by Ruku1590** [comment]
When can we expect this to be available?

**Comment by runev** [comment]
+1 

It's especially useful if you're testing against an Open Banking APIs in Europe which follows Berlin Group's XS2A Framework Implementation Guidelines, which pretty much only allows HTTP message signing to use asymmetric encryption (e.g. rsa-sha256 or rsa-sha512) using the approach described by [https://datatracker.ietf.org/doc/draft-cavage-http-signatures/](draft-cavage-http-signatures-10). 

I've been able to do the entire signing process in a collection's pre-request script, but only using algorithms such as hmac-sha256, because I'm limited to those supported by CryptoJS.

**Comment by runev** [comment]
+1

**Comment by runev** [comment]
+1

**Comment by speaud** [comment]
I agree that it'd be great to integrate NodeJS and it's packages with PM.

In this case you could possibly use the builtin sandbox crypto-js package - https://www.getpostman.com/docs/v6/postman/scripts/postman_sandbox

**Comment by JagpreetKaur** [comment]
When this will be available? Need to use cryto so many times

**Comment by runev** [comment]
This - or at least something which allows us to create RSA-SHA256 signatures based on draft-cavage-http-signatures-10 has to be added.

We've currently had to develop a minor _reverse proxy app_ in Node, which we use to dispatch all Postman requests through to be able to set the Signature header - only because Postman doesn't support RSA-SHA256 using CryptoJS.

It's a pain.

**Comment by logiiiiiiiii** [comment]
+1

**Comment by hakonhw** [comment]
+1

**Comment by anhthang** [comment]
+1

**Comment by msutormin** [comment]
2 years passed and still no crypto. Sadly it doesn't seem like a high priority feature.

**Comment by BoCsA** [comment]
+1

**Comment by joseviro** [comment]
+1

**Comment by nery** [comment]
+1

**Comment by Lukazar** [comment]
+1

**Comment by thoys** [comment]
+1

**Comment by andrew-property-xyz** [comment]
Need crypto in sandbox. All google cloud APIs require OAuth2 tokens signed using RS256. The workaround of using eval(pm.globals.get('jsrsasign-js')) does not work for me. Completely stumped and now looking at other tools to deliver automated testing of my APIs. Clearly this is not a priority and would love to understand the thinking behind supporting crypto-js when RS256 is so commonly used...

**Comment by arlemi** [comment]
@andrew-property-xyz There are a few options to import external libraries, see:
- [Fetching an external library](https://postman-quick-reference-guide.readthedocs.io/en/latest/libraries.html#custom-libraries)
- [Postman External Require](https://github.com/matt-ball/postman-external-require)

**Comment by andrew-property-xyz** [comment]
Hi Arlemi - thanks for getting back to me. With a little more perseverance, I managed to resolve the issue I was having and now have the 'jsrsasign-js' library imported and working. Whose idea was it to have both an underscore and a hyphen in our language...! ;) (it wasn't quite as simple as that but it was the first hurdle)

**Comment by arlemi** [comment]
Glad you got it working @andrew-property-xyz ! If you have a code sample/collection to share I'm sure people in this thread would also be grateful. 😄

**Comment by secure12** [comment]
4 years passed and still no `crypto` in Postman... nice...

**Comment by d-lach** [comment]
This is a joke.

**Comment by muhaimincs** [comment]
what a joke

**Comment by ga250182** [comment]
When we can expect crypto lib made available in postman

**Comment by a-badalyan** [comment]
21.10.2022 ping

**Comment by pbmzero** [comment]
Not having the crypto library easily importable makes  it hard to consume soap wss-v1.1 spec because the UsernameTokenProfile requires sha-1 encrypt the authentication headers `wsse:Nonce` :   
https://www.oasis-open.org/committees/download.php/13392/wss-v1.1-spec-pr-UsernameTokenProfile-01.htm

The SOAP API vendor I'm consuming actually recommended instead use testing framework rather than postman for this reason

**Comment by mtubul** [comment]
any updates reagrding the ability to load crypto module? need it since i am using WebAuthn module which using the crypto module under the hood and i was highly happy if we had it

**Comment by karelbilek** [comment]
Related to #2678 but honestly I don't think it will ever be fixed

**Comment by karelbilek** [comment]
What did work for me:

* copy-paste the minified version of https://kjur.github.io/jsrsasign/ to the pre-request script ( from https://raw.githubusercontent.com/kjur/jsrsasign/master/jsrsasign-all-min.js ) - about 200 lines of minified code
* on top of the script, write `var navigator = {}; var window={}` to satisfy some random checks in the code

then you can actually use the API of the library, as described here

https://kjur.github.io/jsrsasign/

I am not sure if the API you want is there though; maybe you can find another library

**Comment by khiemsens** [comment]
Need to use "crypto" build-in node in Sandbox Postman to create signature authentication for APIs. Have any suggestions for replacement solutions?

**Comment by khiem20tc** [comment]
Need to use "crypto" build-in node in Sandbox Postman to create signature authentication for APIs. Have any suggestions for replacement solutions?

**Comment by lachlanritchie1** [comment]
@davidschwab how did you create the external microservice you mentioned here?  If you can point us to us some exmaples that would be great.  Thank you.


> @harryi3t @kamalaknn Postman/Newman is an awesome platform, but this too is a blocker for RSA support. Exposing the native node.js crypto module would solve, any ETA?
> 
> FYI - [this issue looks similar](https://github.com/postmanlabs/postman-app-support/issues/2678).
> 
> Update: In case anyone else can do the same, I worked around this by creating an external microservice to create the needed signed JWT using the RS256 algorithm.

**Comment by YannMathieu** [comment]
@khiem20tc Take a look at https://gist.github.com/aroemen/91d4368cf5714f5d32ea5bdb53d221f9. I have been able to use it for Auth0 Private Key JWT machine to machine authentication.

**Comment by foliv57** [comment]
Hi, any news on that feature request ? "crypto-js" is now discontinued, meaning that having access to NodeJS crypto module becomes a must-have for any modern encryption. Thank you

**Comment by jbauerrfid** [comment]
When adding `crypto` from NodeJS, we also need full support of `Buffer` (that also allows to get rid of the base64 stuff from crypto-js).
The `Buffer` module allows many byte array manipulation operations, and the NodeJS implementation in v8 has much better performance than crypto-js.

**Comment by appurva21** [comment]
Hey everyone! 👋

Postman's script execution environment now supports the standard web [crypto](https://developer.mozilla.org/en-US/docs/Web/API/Window/crypto) object. Here is an example of signing a payload using ES256:

![Image](https://github.com/user-attachments/assets/ca2a9b96-6998-4ac0-97b4-dff499907aa4)

Not just that, you can now also use external packages from NPM directly into your scripts and quickly get off the blocks using standard packages. Here’s a quick example for generating an ES256 JWT using [jose](https://www.npmjs.com/package/jose):

```js
const { importPKCS8, SignJWT } = pm.require("npm:jose");

const privateKeyPem = `
-----BEGIN PRIVATE KEY-----
...
-----END PRIVATE KEY-----
`;

const key = await importPKCS8(privateKeyPem.trim(), "ES256");

const jwt = await new SignJWT({ foo: "bar" })
  .setProtectedHeader({ alg: "ES256" })
  .setIssuedAt()
  .setExpirationTime("2h")
  .sign(key);

console.log("Generated ES256 JWT:", jwt);
```

![Image](https://github.com/user-attachments/assets/2d3139db-9a37-43b6-b3fe-719c4145dac3)

📌 You can find more details about external packages support and limitations here: https://github.com/postmanlabs/postman-app-support/issues/13542


I’ll go ahead and close this one for now, but feel free to open a new ticket if you have more questions or run into any issues. 🚀

**Comment by msutormin** [comment]
That's great! But jeez, it took almost 8 years to close the ticket.

**Comment by DSchau** [comment]
@msutormin better late than never! Let us know what you think of the new functionality and how we can keep improving it. We'll try to dramatically improve that 8-year resolution time!
